### PR TITLE
safer remove_bos logic

### PR DESCRIPTION
**Pull request by andyrdt**
The `remove_bos` logic in `ActivationBuffer` assumes right padding (i.e. it assumes that `bos` will always appear as the left-most token):

```
            if self.remove_bos:
                hidden_states = hidden_states[:, 1:, :]
                attn_mask = attn_mask[:, 1:]
```

However, it looks like [by default `nnsight` uses left padding](https://github.com/ndif-team/nnsight/blob/12d436c1accf27386bd79631551a8b6677b57dc1/src/nnsight/modeling/language.py#L119).

This means that unless a user explicitly sets the padding side to right, they will train on `bos` (if using the nnsight-based ActivationBuffer).

I got bit by this unfortunately, and my evaluation metrics (in particular fraction of variance explained) look quite different.

I've added asserts surface this assumption. Could alternatively modify the tokenizer's padding side directly, but I think surfacing the error seems better.

### CODE REVIEW

**Comment by andyrdt** [comment]
After chatting with Adam, decided to filter bos tokens by applying a mask to the activations. This solution seems better:
- robust to `padding_side`
- compatible with sequence packing

```python
            mask = (input["attention_mask"] != 0)
            if self.remove_bos:
                bos_mask = (input["input_ids"] == self.tokenizer.bos_token_id)
                mask = mask & ~bos_mask
            hidden_states = hidden_states[mask]
```

**Comment by adamkarvonen** [comment]
Thanks!
